core/drivers/crypto/hisilicon/sec_pbkdf2.c

```diff
@@ -0,0 +1,348 @@
+	}
+
+	if (salt_len > SEC_MAX_SALT_LEN) {
+		EMSG("Unsupported salt len %lu", salt_len);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return sec_pbkdf2_dk_iteration_check(alg, c_num, derived_key_len);
+}
+
+static TEE_Result sec_pbkdf2_set_derive_type(uint32_t hash_id,
+					     struct sec_pbkdf2_msg *msg)
+{
+	uint32_t alg = TEE_ALG_HMAC_ALGO(hash_id);
+
+	switch (alg) {
+	case TEE_ALG_HMAC_SHA1:
+		msg->derive_type = SEC_HMAC_SHA1;
+		break;
+	case TEE_ALG_HMAC_SHA224:
+		msg->derive_type = SEC_HMAC_SHA224;
+		break;
+	case TEE_ALG_HMAC_SHA256:
+		msg->derive_type = SEC_HMAC_SHA256;
+		break;
+	case TEE_ALG_HMAC_SHA384:
+		msg->derive_type = SEC_HMAC_SHA384;
+		break;
+	case TEE_ALG_HMAC_SHA512:
+		msg->derive_type = SEC_HMAC_SHA512;
+		break;
+	case TEE_ALG_HMAC_SM3:
+		msg->derive_type = SEC_HMAC_SM3;
+		break;
+	default:
+		EMSG("Invalid hamc alg type 0x%x", alg);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_pbkdf2_set_buf(const uint8_t *password,
+				     const uint8_t *salt,
+				     struct sec_pbkdf2_msg *msg)
+{
+	msg->key_dma = virt_to_phys(msg->base_key);
+	if (!msg->key_dma) {
+		EMSG("Fail to get key dma addr");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	msg->salt_dma = virt_to_phys(msg->salt);
+	if (!msg->salt_dma) {
+		EMSG("Fail to get salt dma addr");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	msg->out_dma = virt_to_phys(msg->out);
+	if (!msg->out_dma) {
+		EMSG("Fail to get out dma addr");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (password)
+		memcpy(msg->base_key, password, msg->key_len);
+
+	memcpy(msg->salt, salt, msg->salt_len);
+
+	return TEE_SUCCESS;
+}
+
+static void sec_pbkdf2_clean_buf(struct sec_pbkdf2_msg *msg)
+{
+	memzero_explicit(msg->base_key, msg->key_len);
+	memzero_explicit(msg->salt, msg->salt_len);
+}
+
+static TEE_Result sec_pbkdf2_msg_init(uint32_t hash_id, size_t password_len,
+				      size_t salt_len, size_t derived_key_len,
+				      uint32_t iteration_count,
+				      struct sec_pbkdf2_msg *msg)
+{
+	msg->key_len = password_len;
+	msg->salt_len = salt_len;
+	msg->out_len = derived_key_len;
+	msg->c_num = iteration_count;
+	return sec_pbkdf2_set_derive_type(hash_id, msg);
+}
+
+TEE_Result tee_cryp_pbkdf2(uint32_t hash_id, const uint8_t *password,
+			   size_t password_len, const uint8_t *salt,
+			   size_t salt_len, uint32_t iteration_count,
+			   uint8_t *derived_key, size_t derived_key_len)
+{
+	struct sec_pbkdf2_msg *msg = NULL;
+	TEE_Result ret;
+
+#ifndef CFG_CRYPTO_HW_PBKDF2_WITH_EFUSE
+	if (!password) {
+		EMSG("Password buf is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+#endif
+	if (!salt || !derived_key) {
+		EMSG("Invalid pbkdf2 buf");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = sec_pbkdf2_params_check(hash_id, password_len, salt_len,
+				      iteration_count, derived_key_len);
+	if (ret)
+		return ret;
+
+	msg = malloc(sizeof(*msg));
+	if (!msg) {
+		EMSG("Fail to malloc msg");
+		return TEE_ERROR_STORAGE_NO_SPACE;
```

Comment:
`TEE_ERROR_OUT_OF_MEMORY`
FYI `TEE_ERROR_STORAGE_NO_SPACE` relates to the secure storage space.